Extend validate-modules to also validate plugins

felixfontein opened the pull request:
##### SUMMARY
Adds a `--plugin-type` option to `validate-modules`, and modifies ansible-test to use it.

~~There are a lot of things left to standardize / decide, but this is a first attempt to see how it looks :)~~

~~It includes #71679 and #71735 since I think these can get merged soon, and I didn't want to have conflicts with them or wait for them...~~

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ansible-test validate-modules

felixfontein committed:
Let validate-modules also validate plugins.

felixfontein commented:
To discuss:

- what's with `name:` next to `description:` in option descriptions? (see default_callback docs_fragment for extensive usage)
- plugin types which must have / can have / cannot have examples
- author: required with same conditions as for modules, or optional, or other conditions?
- `<plugin_type>: <name>` vs. `plugin_type: <plugin_type>` and `name: <name>`
- option type: should long forms `boolean`, `string`, `integer` etc. be allowed?
- return value `type`: required or not for plugins?

felixfontein commented:
Another thing that can be discussed:
 - for callback plugins, `type` vs. `callback_type` - @abadger chose `type` in antsibull-docs, but `callback_type` seems to be very common as well (and is named more closely to the `CALLBACK_TYPE` field in the plugin itself)

gundalow commented on test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py:
```diff
@@ -2281,8 +2315,8 @@ def re_compile(value):
 
 def run():
     parser = argparse.ArgumentParser(prog="validate-modules")
-    parser.add_argument('modules', nargs='+',
```
I'm not sure if anyone has ever used these options. 
Do we need to keep an (undocumented) `--modules` for backwards compatability?

felixfontein replied:
`modules` is in fact just the variable the positional arguments are stored in (and maybe it's also used as the default placeholder string in the help), so the CLI interface is already backwards compatible.

gundalow commented on test/lib/ansible_test/_internal/sanity/validate_modules.py:
```diff
@@ -48,6 +50,39 @@
 
 
+# from ansible.constants import DOCUMENTABLE_PLUGINS
+DOCUMENTABLE_PLUGINS = (
```
Could this opt-in (rather than opt-out) mean we don't validate documentation if there is a new plugin type on the future?

felixfontein replied:
Yes. I originally wanted to import `DOCUMENTABLE_PLUGINS` from `ansible.constants`, but that seems to be somewhere between hard to impossible. I tried something similar to https://github.com/ansible/ansible/blob/devel/test/lib/ansible_test/_internal/util.py#L218-L234 but unfortunately `constants` depends on a large chunk of Ansible. @mattclay do you have an idea how to solve this in a good way?

mattclay replied:
We have similar lists in other places as well:

https://github.com/ansible/ansible/blob/fc08c1f3c554c9cd4ad39f846d123408de19ad2a/docs/docsite/Makefile#L35

https://github.com/ansible/ansible/blob/fc08c1f3c554c9cd4ad39f846d123408de19ad2a/test/lib/ansible_test/_internal/sanity/ansible_doc.py#L52-L61

There are other locations that also list plugins types, mostly for non-documentation purposes. I don't think we'll be unifying all the various lists anytime soon.

@felixfontein What's the issue with just importing `ansible.constants` directly from within validate-modules itself?

felixfontein replied:
@mattclay then validate-modules itself needs to figure out which path belongs to which plugin class. That's doable, but it is a lot easier to do in ansible-test directly since it already knows the paths (`data_context().content.plugin_paths`).

mattclay replied:
@felixfontein We could have ansible-test pass the plugin list and paths to validate-modules when invoking it.

felixfontein replied:
@mattclay true, but I think it would make validate-modules more complicated (also more complicated to use stand-alone, though I don't know if anyone actually does that).

What do you prefer?

felixfontein replied:
I yesterday learned that the ansible-doc sanity test has a similar hard-coded list of plugin types that are not documentable. I've moved that list to a common constant also use it in here now, so we have only one place left where (un-)documentable plugins are listed.

felixfontein replied:
(See 75c4ffb0803e9f20268ca7cd975bce8922d19548)

gundalow reviewed (commented):
Amazing. Thank you so much for this.
Only done a quick skim so far. 

I wonder if fixing the new the documentation problems this finds is something that people could help with in the next Hackathon?

felixfontein commented:
I think almost all of the problems shown are fixed in #71736. The only exception is probably the `paramiko` name discrepancy, which needs some special casing in this PR. I will add that once #71736 has been merged.

felixfontein commented:
I rebased to remove the commits from #71679 since that has been merged by now.

felixfontein commented:
Rebasing after #71736 has been merged. Should resolve most errors.

mattclay commented:
Something else to consider is that the `import` sanity test should be extended to cover plugins as well, since `validate-modules` will need to import them.

We'll also need to document the expected behavior for plugins under test. For example, plugins will need to be importable with only ansible's own dependencies available, similar to how modules are required to be importable with only the standard library modules available.

Some plugins will need to be updated to accommodate this, and we should document the recommended approach.

felixfontein commented:
Rebased to remove conflict.

felixfontein commented:
I've rebased and adjusted to the new `cli:` type for plugin options (only used by the `ssh` plugin apparently). I also tested this with a couple of my collections, and they all passed (resp. were made passing in ansible-collections/community.general#2414).

ready_for_review

felixfontein committed:
Support 'option' in 'cli'.

felixfontein commented:
Rebased to latest `devel`, and checked against a set of collections as well (resulting in https://github.com/ansible-collections/community.dns/pull/67).

ready_for_review

Andersson007 commented:
Would be nice to see plugin doc validation merged as the documentation is broken currently in some collections, for example

![Screenshot from 2021-09-28 11-13-28](https://user-images.githubusercontent.com/34477873/135059112-7d675e63-c760-4b42-9d25-6f4911932e71.png)

bcoca commented on test/lib/ansible_test/_internal/commands/sanity/__init__.py:
```diff
@@ -123,6 +123,21 @@
+    # (https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#modules--plugins)
+    'plugin_utils',
+    'sub_plugins',
+])
```
I would not create new constant but reuse the DOCUMENTABLE_PLUGINS or CONFIGURABLE_PLUGINS ones from ansible itself. FYI, action plugins are documented via modules of the same name, see 'template'.

felixfontein replied:
Ref: https://github.com/ansible/ansible/pull/71734#discussion_r487525585 from over a year ago :)

I'll try again to figure out whether it works better now.

> FYI, action plugins are documented via modules of the same name, see 'template'.

That's why action plugins must be listed here, since they themselves do not contain documentation.

felixfontein replied:
When trying to import this when running the sanity tests inside the default Docker container, I get:
```
[...]
  File "/root/ansible/test/lib/ansible_test/_internal/commands/sanity/ansible_doc.py", line 9, in <module>
    from ansible.constants import DOCUMENTABLE_PLUGINS
ModuleNotFoundError: No module named 'ansible'
```
While the workaround suggested by @mattclay in https://github.com/ansible/ansible/pull/71734#discussion_r490513071 would work for validate_modules, it doesn't work for the ansible-doc sanity test (if I understood it correctly).

If this is the only problem with this PR, my suggestion would be to keep the list (because this PR just moves it from one place to another in ansible-test, and starts using that list in another place).

bcoca replied:
my issue is that the lists will fall out of sync and one is additive while the other is subtractive ... fun all around in future.

bcoca replied:
if direct import does not work, maybe a query to ansible-doc to get 'supported' types is in order?

felixfontein replied:
I did rename the list to `DOCUMENTABLE_PLUGINS` and added a note to ansible.constants.DOCUMENTABLE_PLUGINS that also this list needs to be updated. That way it should be easier to keep them in sync.

felixfontein committed:
Use DOCUMENTABLE_PLUGINS instead of UNDOCUMENTED_PLUGIN_TYPES.

mkrizek commented:
!waffling stale_ci

felixfontein committed:
module -> modules.

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/ansible_doc.py:
```diff
@@ -6,6 +6,8 @@
 import re
 import typing as t
 
+from . import DOCUMENTABLE_PLUGINS
```
This should be added to the existing `from . import` list below.

felixfontein replied:
Addressed in accfc7f632680fefa41206deb040e67b2d879400.

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -5,6 +5,10 @@
 
+from collections import defaultdict
+
+from . import DOCUMENTABLE_PLUGINS
```
This should be added to the existing `from . import` list below.

felixfontein replied:
Addressed in accfc7f632680fefa41206deb040e67b2d879400.

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -5,6 +5,10 @@
 import os
 import typing as t
 
+from collections import defaultdict
```
Import `collections` and then use `collections.defaultdict` instead. That will match the code style used throughout ansible-test.

felixfontein replied:
Addressed in accfc7f632680fefa41206deb040e67b2d879400.

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/__init__.py:
```diff
@@ -123,6 +123,11 @@
 
+# NOTE: must match ansible.constants.DOCUMENTABLE_PLUGINS, but with 'module' replaced by 'modules'!
+DOCUMENTABLE_PLUGINS = (
+    'become', 'cache', 'callback', 'cliconf', 'connection', 'httpapi', 'inventory', 'lookup', 'netconf', 'shell', 'vars', 'modules', 'strategy'
```
It would be good to keep the plugin list sorted:

```suggestion
    'become', 'cache', 'callback', 'cliconf', 'connection', 'httpapi', 'inventory', 'lookup', 'modules', 'netconf', 'shell', 'strategy', 'vars'
```

felixfontein replied:
My goal was to keep the list as close to the one in `ansible.constants`. But I can also sort it.

felixfontein replied:
Addressed in accfc7f632680fefa41206deb040e67b2d879400.

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -54,6 +58,32 @@
 )
 
 
+def _get_plugin_type_getter():
```
Move this initialization into `ValidateModulesTest.__init__` and make `get_plugin_type` a class method.

felixfontein replied:
Addressed in accfc7f632680fefa41206deb040e67b2d879400.

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -54,6 +58,32 @@
+    if not data_context().content.collection:
+        exceptions.add('lib/ansible/plugins/cache/base.py')
+
+    def get_plugin_type(target):  # type: TestTarget -> t.Optional[str]
```
Add a docstring for this function.

Also, the type hint needs to be corrected:

```suggestion
    def get_plugin_type(target):  # type: (TestTarget) -> t.Optional[str]
```

felixfontein replied:
Addressed in accfc7f632680fefa41206deb040e67b2d879400.

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -54,6 +58,32 @@
+    prefixes = dict(
+        (plugin_type, plugin_path + '/')
+        for plugin_type, plugin_path in content.plugin_paths.items()
+        if plugin_type != 'module' and plugin_type in DOCUMENTABLE_PLUGINS
```
This was probably intended to be `'modules'` since that's what is in `content.plugin_paths`.

felixfontein replied:
Addressed in accfc7f632680fefa41206deb040e67b2d879400.

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -54,6 +58,32 @@
+    )
+    exceptions = set()
+    for prefix in prefixes.values():
+        exceptions.add(prefix + '__init__.py')
```
Use `os.path.join` instead of string concatenation.

felixfontein replied:
Addressed in accfc7f632680fefa41206deb040e67b2d879400.

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -54,6 +58,32 @@
+        for plugin_type, plugin_path in content.plugin_paths.items()
+        if plugin_type != 'module' and plugin_type in DOCUMENTABLE_PLUGINS
+    )
+    exceptions = set()
```
This could be a set comprehension. Also, I'd call it `exclusions` instead of `exceptions`.

felixfontein replied:
Addressed in accfc7f632680fefa41206deb040e67b2d879400.

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -54,6 +58,32 @@
 
 
+def _get_plugin_type_getter():
+    content = data_context().content
```
No need to put this into a variable since it's used only once.

felixfontein replied:
It is used in a loop though, which will then cause `data_context()` to be called once per loop iteration. Should I still change it?

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -54,6 +58,32 @@
 
+def _get_plugin_type_getter():
+    content = data_context().content
+    prefixes = dict(
```
This can be a dict comprehension.

felixfontein replied:
Addressed in accfc7f632680fefa41206deb040e67b2d879400.

felixfontein committed:
Address review comments.

mkrizek commented:
!waffling core_review

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -64,28 +66,51 @@ def __init__(self):
+        content = data_context().content
+        self._prefixes = {
+            plugin_type: plugin_path + '/'
+            for plugin_type, plugin_path in content.plugin_paths.items()
```
A separate variable isn't needed since `data_context()` is only resolved once for this comprehension:

```suggestion
        self._prefixes = {
            plugin_type: plugin_path + '/'
            for plugin_type, plugin_path in data_context().content.plugin_paths.items()
```

felixfontein replied:
Fixed in 3673f6055ce08b11653dae65a76a35b84fe0f555.

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -64,28 +66,51 @@ def __init__(self):
+            for plugin_type, plugin_path in content.plugin_paths.items()
+            if plugin_type != 'modules' and plugin_type in DOCUMENTABLE_PLUGINS
+        }
+        self._exclusions = {os.path.join(prefix, '__init__.py') for prefix in self._prefixes.values()}
```
Add blank lines for readability:

```suggestion

        self._exclusions = {os.path.join(prefix, '__init__.py') for prefix in self._prefixes.values()}

```

felixfontein replied:
Fixed in 3673f6055ce08b11653dae65a76a35b84fe0f555 (I also added a few more than you suggested).

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -64,28 +66,51 @@ def __init__(self):
         return 'A100'
 
+    def get_plugin_type(self, target):  # type: (TestTarget) -> t.Optional[str]
+        """Return the plugin type of the given target, or None if it is neither a plugin or module."""
```
```suggestion
        """Return the plugin type of the given target, or None if it is not a plugin or module."""
```

felixfontein replied:
Fixed in 3673f6055ce08b11653dae65a76a35b84fe0f555.

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -64,28 +66,51 @@ def __init__(self):
+        for plugin_type, prefix in self._prefixes.items():
+            if target.path.startswith(prefix):
+                return plugin_type
+        return None
```
```suggestion

        return None
```

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -64,28 +66,51 @@ def __init__(self):
         settings = self.load_processor(args)
 
-        paths = [target.path for target in targets.include]
+        target_per_type = collections.defaultdict(list)
```
```suggestion
        target_per_type = collections.defaultdict(list)

```

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -109,37 +134,48 @@ def test(self, args, targets, python):  # type: (SanityConfig, SanityTargets, Py
-            stdout = ex.stdout
-            stderr = ex.stderr
-            status = ex.status
+        errors = []
```
```suggestion
        errors = []

```

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -109,37 +134,48 @@ def test(self, args, targets, python):  # type: (SanityConfig, SanityTargets, Py
-        errors = []
+            messages = json.loads(stdout)
+
+            plugin_errors = []
```
```suggestion
            plugin_errors = []

```

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -109,37 +134,48 @@ def test(self, args, targets, python):  # type: (SanityConfig, SanityTargets, Py
-                    code='%s' % item['code'],
-                    message=item['msg'],
-                ))
+            plugin_errors = settings.process_errors(plugin_errors, paths)
+            errors += plugin_errors
```
```suggestion
            errors += settings.process_errors(plugin_errors, paths)
```

felixfontein replied:
Fixed in 3673f6055ce08b11653dae65a76a35b84fe0f555.

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -64,28 +66,51 @@ def __init__(self):
+        self._prefixes = {
+            plugin_type: plugin_path + '/'
+            for plugin_type, plugin_path in content.plugin_paths.items()
+            if plugin_type != 'modules' and plugin_type in DOCUMENTABLE_PLUGINS
```
Looking at this closer, it looks like this can be simplified to:

```suggestion
            if plugin_type in DOCUMENTABLE_PLUGINS
```

Since the only reason to exclude `'modules'` is to later report `'module'` instead -- but then that value is never used outside of a conditional, which could simply be updated to match.

felixfontein replied:
I think this simplification has some unintended side-effects, since `__init__.py` files in subdirectories in `plugins/modules/` will now be matched as "modules".

felixfontein replied:
Ok, I tried it a bit different in ccf9b5c5673b000f2a7b80e024c01ee8cf3a824f, by adding an explicit check for ending with `/__init__.py`. With that it should work.

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -64,28 +66,51 @@ def __init__(self):
+        """Return the plugin type of the given target, or None if it is neither a plugin or module."""
+        if target.path in self._exclusions:
+            return None
+        if target.module:
+            return 'module'
```
```suggestion

```

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -109,37 +134,48 @@ def test(self, args, targets, python):  # type: (SanityConfig, SanityTargets, Py
+        for plugin_type, plugin_targets in sorted(target_per_type.items()):
+            paths = [target.path for target in plugin_targets]
+            plugin_cmd = list(cmd)
+            if plugin_type != 'module':
+                plugin_cmd += ['--plugin-type', plugin_type]
```
```suggestion

            if plugin_type != 'modules':
                plugin_cmd += ['--plugin-type', plugin_type]

```

felixfontein replied:
Fixed in 3673f6055ce08b11653dae65a76a35b84fe0f555.

felixfontein committed:
Addressing more review comments.

felixfontein committed:
More newlines.

felixfontein committed:
Avoid special case for modules.

mkrizek commented:
!waffling needs_revision

felixfontein committed:
Support 'keyword', clean up error codes.

felixfontein commented:
I've updated the PR to add support for `keyword` (864c2591b6f0861321009e89462d1baaef9ea39a).

mattclay commented on test/lib/ansible_test/_internal/commands/sanity/validate_modules.py:
```diff
@@ -109,37 +139,51 @@ def test(self, args, targets, python):  # type: (SanityConfig, SanityTargets, Py
+                        message=item['msg'],
+                    ))
+
+            errors += settings.process_errors(plugin_errors, paths)
```
Call this once outside the loop. It just needs the list of all the errors and all the paths tested.

felixfontein replied:
I moved it outside the loop in 74682b3e1af3db3d02d4219c8905a5144be99269.

mattclay commented on test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/__init__.py:
```diff
@@ -17,4 +17,4 @@
 from __future__ import (absolute_import, division, print_function)
 __metaclass__ = type
-__version__ = '0.0.1b'
+__version__ = '0.1.0'
```
I don't think this is used anywhere, so it can be removed.

felixfontein replied:
I removed it in 74682b3e1af3db3d02d4219c8905a5144be99269.

felixfontein committed:
Call settings.process_errors only once; remove __version__.

s-hertel reviewed (approved):
@felixfontein could you add a changelog?

felixfontein committed:
Add changelog fragment.

felixfontein commented:
Looks like I completely forgot about that one :laughing: Now it's there, thanks!
